rkt/api_service.go

```diff
@@ -691,8 +695,76 @@ func (s *v1AlphaAPIServer) InspectImage(ctx context.Context, request *v1alpha.In
 	return &v1alpha.InspectImageResponse{Image: image}, nil
 }
 
+// LogsStreamWriter is a wrapper around a GRPC streaming server.
+// Implements io.Writer interface.
+type LogsStreamWriter struct {
+	server v1alpha.PublicAPI_GetLogsServer
+}
+
+func (sw LogsStreamWriter) Write(b []byte) (int, error) {
+	if err := sw.server.SendMsg(b); err != nil {
+		return 0, err
+	}
+	return len(b), nil
+}
+
 func (s *v1AlphaAPIServer) GetLogs(request *v1alpha.GetLogsRequest, server v1alpha.PublicAPI_GetLogsServer) error {
-	return fmt.Errorf("not implemented yet")
+	uuid, err := types.NewUUID(request.PodId)
+	if err != nil {
+		return err
+	}
+	pod, err := getPod(uuid)
+	if err != nil {
+		return err
+	}
+	stage1Path := "stage1/rootfs"
+	if pod.usesOverlay() {
+		stage1TreeStoreID, err := pod.getStage1TreeStoreID()
+		if err != nil {
+			return err
+		}
+		stage1Path = fmt.Sprintf("/overlay/%s/upper/", stage1TreeStoreID)
+	}
+	path := filepath.Join(getDataDir(), "/pods/run/", request.PodId, stage1Path, "/var/log/journal/")
+	jconf := sdjournal.JournalReaderConfig{
+		Path: path,
+	}
+	if request.AppName != "" {
+		acname, err := types.NewACName(request.AppName)
+		if err != nil {
+			return err
+		}
+		jconf.Matches = []sdjournal.Match{
+			{
+				Field: sdjournal.SD_JOURNAL_FIELD_SYSTEMD_UNIT,
+				Value: stage1common.ServiceUnitName(*acname),
+			},
+		}
+	}
+	if request.SinceTime != 0 {
+		t := time.Unix(request.SinceTime, 0)
+		jconf.Since = time.Since(t)
+	}
+	if request.Lines != 0 {
+		jconf.NumFromTail = uint64(request.Lines)
+	}
+
+	jr, err := sdjournal.NewJournalReader(jconf)
```

Comment:
Oh actually it's not quite that bad because of https://github.com/coreos/go-systemd/pull/141#issuecomment-179221840 . But we still shouldn't be opening the log every time if we can avoid it?